packages/components/pie-input/src/input.scss

```diff
@@ -57,6 +71,31 @@
             margin-inline-start: 0;
         }
     }
+
+    &:hover {
+        --input-container-clr: hsl(var(--dt-color-container-default-h), var(--dt-color-container-default-s), calc(var(--dt-color-container-default-l) + calc(-1 * var(--dt-color-hover-01))));
+    }
+
+    &[data-pie-readonly] {
+        --input-container-clr: var(--dt-color-disabled-01);
+        --input-border-clr: var(--dt-color-border-inverse);
+    }
+
+    // Ensure that if an input is readonly and disabled, the disabled styles take precedence
+    &[data-pie-disabled][data-pie-readonly],
+    &[data-pie-disabled]{
+        --input-container-clr: var(--dt-color-disabled-01);
+        --input-border-clr: var(--dt-color-disabled-01);
+        --input-text-clr: var(--dt-color-content-disabled);
+        --input-text-clr-placeholder: var(--dt-color-content-disabled);
+        --input-text-clr-leading-trailing-content: var(--dt-color-content-disabled);
+    }
+
+    &:focus-within {
+        outline: var(--input-outline-width) solid var(--dt-color-focus-outer);
+        outline-offset: var(--input-outline-width);
+        box-shadow: 0 0 0 var(--input-outline-width) var(--dt-color-focus-inner);
+    }
```

Comment:
would love to discuss this approach. it seems suspiciously simple but appears to work quite well

Reply:
one thing I notice is outline appears to ignore border radius on ios safari 14. However it doesn't look terrible and could be a form of progressive enhancement if that is the only issue

<img width="297" alt="Screenshot 2024-04-16 at 21 23 48" src="https://github.com/justeattakeaway/pie/assets/16766185/10fd7df2-1dfa-4549-bcb6-01894f2d61e9">

Reply:
I think if it's only like that in v14 and resolves in later versions, then that's absolutely fine 👍